src/Extract.hs

```diff
@@ -73,23 +73,28 @@ locsFromExp = nub . go
       IntEnv _ -> []
       ByEnv _ -> []
       ITE x y z -> go x <> go y <> go z
-      TEntry a -> case a of
-        DirectInt contract name -> [IntLoc $ DirectInt contract name]
-        DirectBool contract slot -> [BoolLoc $ DirectBool contract slot]
-        DirectBytes contract slot -> [BytesLoc $ DirectBytes contract slot]
-        MappedInt contract name ixs -> [IntLoc $ MappedInt contract name ixs] <> ixLocs ixs
-        MappedBool contract name ixs -> [BoolLoc $ MappedBool contract name ixs] <> ixLocs ixs
-        MappedBytes contract name ixs -> [BytesLoc $ MappedBytes contract name ixs] <> ixLocs ixs
-        where
-          ixLocs :: NonEmpty.NonEmpty ReturnExp -> [StorageLocation]
-          ixLocs = concatMap locsFromReturnExp
+      TEntry a _ -> locsFromStorageItem a
+
+    locsFromStorageItem :: TStorageItem t a -> [StorageLocation]
+    locsFromStorageItem t = case t of
+      IntItem contract name ixs -> [IntLoc $ IntItem contract name $ untimeTyped <$> ixs] <> ixLocs ixs
+      BoolItem contract name ixs -> [BoolLoc $ BoolItem contract name $ untimeTyped <$> ixs] <> ixLocs ixs
+      BytesItem contract name ixs -> [BytesLoc $ BytesItem contract name $ untimeTyped <$> ixs] <> ixLocs ixs
+      where
+        ixLocs :: [TypedExp t] -> [StorageLocation]
+        ixLocs = concatMap (locsFromTypedExp . untimeTyped)
+
+        untimeTyped :: TypedExp t -> TypedExp Untimed
```

Comment:
I'm guessing you mean `untimeTyped`? Because a `StorageLocation` always contains an untimed `TStorageItem`, but we don't know the timing of the input to `locsFromExp`. So the `TStorageItem` may be timed, which means that its indices will be timed as well. We need to remove that. This is the sole valid use of `forceTime` that I can think of.

(And yes, `locsFromExp` has to be polymorphic in the timing, because it's used in `locsFromBehaviour` and `locsFromConstructor` to calculate all locations, both from timed and untimed expressions.)

Reply:
ah I see, and presumably if we modified `StorageLocation` as below to expose the timing in its type we would end up with a bunch of issues elsewhere?

```haskell
data StorageLocation (t :: Timing) where
  IntLoc :: t -> StorageLocation (TStorageItem t Integer)
  BoolLoc :: t -> StorageLocation (TStorageItem t Bool)
  BytesLoc :: t -> StorageLocation (TStorageItem t Bool)
deriving instance Show (StorageLocation t)
deriving instance Eq (StorageLocation t)
```

Reply:
or wait the above doesn't even compile hehe :woman_facepalming:

Reply:
It's a good question, I toyed with that solution myself. What you want is probably this though:

```haskell
data StorageLocation (t :: Timing) where
  IntLoc :: TStorageItem t Integer -> StorageLocation t
  BoolLoc :: TStorageItem t Bool -> StorageLocation t
  BytesLoc :: TStorageItem t ByteString -> StorageLocation t
deriving instance Show (StorageLocation t)
deriving instance Eq (StorageLocation t)
```

And yes, this would give us a bunch of issues elsewhere. For starters. we couldn't implement `locsFromBehaviour :: Behaviour -> [StorageLocation t]` because `t` would not be uniform across all members of the list. Additionally, to me it doesn't really make sense to say that a *location* is timed. A location just exists, but we can reference it at different times.